lighthouse-core/audits/robust-csp.js

```diff
@@ -0,0 +1,188 @@
+/**
+ * @license Copyright 2021 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Audit = require('./audit.js');
+const MainResource = require('../computed/main-resource.js');
+const i18n = require('../lib/i18n/i18n.js');
+const {
+  evaluateRawCspForFailures,
+  evaluateRawCspForWarnings,
+  evaluateRawCspForSyntax,
+  getTranslatedDescription,
+} = require('../lib/csp-evaluator.js');
+
+/** @typedef {import('../lib/csp-evaluator.js').Finding} Finding */
+
+const UIStrings = {
+  /** Title of a Lighthouse audit that evaluates the security of a page's CSP. This descriptive title is shown to users when no security vulnerabilities are found in the CSP. "CSP" stands for "Content Security Policy". "XSS" stands for "Cross Site Scripting". "CSP" and "XSS" do not need to be translated. */
+  title: 'CSP is robust against XSS attacks',
```

Comment:
There's also clickjacking and mixed-content mistakes, which I believe this audit will also surface.

https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/45542.pdf section 2.2

Reply:
> • Clickjacking: forcing users to take unwanted actions
in an affected application by overlaying hidden frames
on attacker-controlled pages (protected by restricting
framing with frame-ancestors)
• Mixed content: Accidentally loading resources from insecure protocols on pages delivered over HTTPS (protected with the upgrade-insecure-requests and blockall-mixed-content keywords and by restricting the
loading of scripts and sensitive resources to https:).

These audit does not suggest anything about `frame-ancestors`, `upgrade-insecure-requests`, or `blockall-mixed-content` so it would not surface vulnerabilities to clickjacking and mixed content.

Reply:
Then perhaps `robust-csp` is not the best name for this audit. If we want to separate these concerns, we could name this `csp-xss`.